[Python] Improve Python interop usability.

dan-zheng opened the pull request:
- Define `version` and `versionInfo` computed properties on `PythonInterface`
  for convenience.
- Define labelless failable initializers from `PythonObject`
  (representing `numpy.ndarray`) to `ShapedArray` and `Tensor`.
  - A labelless initializer was not created for `Array` because `Array` is
    always one-dimensional, so the nature of the conversion is semantically bigger.

dan-zheng committed:
[Python] Improve Python interop usability.

- Define `version` and `versionInfo` computed properties on `PythonInterface`
  for convenience.
- Define labelless failable initializers from `PythonObject`
  (representing `numpy.ndarray`) to `ShapedArray` and `Tensor`.
  - A labelless initializer was not created for `Array` because `Array` is
    always one-dimensional, so the conversion is semantically more significant.
    Users can use the labelled `Array(numpyArray:)` initalizer.

rxwei commented on stdlib/public/Python/Python.swift:
```diff
@@ -684,6 +684,18 @@ public struct PythonInterface {
+
+  // The Python runtime version.
+  // Equivalent to `sys.version` in Python.
+  public var version: PythonObject {
```
Add these to `PythonLibrary` instead of `Python`.

rxwei commented on stdlib/public/TensorFlow/NumpyConversion.swift:
```diff
@@ -109,6 +109,22 @@ extension Tensor : ConvertibleFromNumpyArray
+  // Creates a `ShapedArray` instance from the given `PythonObject` if it is a
+  // `numpy.ndarray` instance with a matching scalar datatype.
+  public init?(_ pythonObject: PythonObject) {
+    self.init(numpyArray: pythonObject)
```
If it's not a numpy array, we should load it as a (potentially multi-dimensional) python array.

dan-zheng replied:
Good point. Added TODO comments in https://github.com/apple/swift/pull/22104/commits/49f70e4e5ee106f45493308a2ce308929e2d20ac.

dan-zheng committed:
Add TODO comments for extending failable initializer logic from `PythonObject`.

dan-zheng committed:
[Python] Rename `numpy.ndarray` initializer to `init?(numpy:)`.

`init?(numpy:)` is a bit shorter than the previous `init?(numpyArray:)`.
Using a labelless initializer is problematic because it causes type inference
problems due to implicit conversions.

rxwei commented:
@swift-ci please test tensorflow Linux

rxwei commented:
@swift-ci please test tensorflow Linux

rxwei commented:
@swift-ci please test tensorflow Linux

rxwei commented:
@swift-ci please test tensorflow Linux

rxwei commented:
@swift-ci please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
